pandas/src/inference.pyx

```diff
@@ -629,48 +626,155 @@ cdef extern from "parse_helper.h":
+    na_values : set
+        Set of values that should be interpreted as NaN.
+    convert_empty : bool, default True
+        If an empty array-like object is encountered, whether to interpret
+        that element as NaN or not. If set to False, a ValueError will be
+        raised if such an element is encountered and 'coerce_numeric' is False.
+    coerce_numeric : bool, default False
+        If initial attempts to convert to numeric have failed, whether to
+        force conversion to numeric via alternative methods or by setting the
+        element to NaN. Otherwise, an Exception will be raised when such an
+        element is encountered.
+
+        This boolean also has an impact on how conversion behaves when a
+        numeric array has no suitable numerical dtype to return (i.e. uint64,
+        int32, uint8). If set to False, the original object array will be
+        returned. Otherwise, a ValueError will be raised.
+
+    Returns
+    -------
+    numeric_array : array of converted object values to numerical ones
     """
     cdef:
         int status, maybe_int
         Py_ssize_t i, n = values.size
         ndarray[float64_t] floats = np.empty(n, dtype='f8')
         ndarray[complex128_t] complexes = np.empty(n, dtype='c16')
         ndarray[int64_t] ints = np.empty(n, dtype='i8')
+        ndarray[uint64_t] uints = np.empty(n, dtype='u8')
         ndarray[uint8_t] bools = np.empty(n, dtype='u1')
+        bint seen_null = False
+        bint seen_uint = False
+        bint seen_sint = False
         bint seen_float = False
         bint seen_complex = False
         bint seen_int = False
         bint seen_bool = False
         object val
         float64_t fval
 
+    def check_uint64_nan():
+        """
+        Check whether we have encountered uint64 when handling a NaN element.
+
+        If uint64 has been encountered, we cannot safely cast to float64 due
+        to truncation problems (this would occur if we return a numeric array
+        containing a NaN element).
+
+        Returns
+        -------
+        return_values : bool
+            Whether or not we should return the original input array to avoid
+            data truncation.
+        """
+        if seen_null and seen_uint:
+            if not coerce_numeric:
+                return True
+            else:
+                raise ValueError("uint64 array detected, and such an "
+                                 "array cannot contain NaN.")
+
+        return False
+
+    def check_uint64_int64_conflict():
+        """
+        Check whether we have encountered both int64 and uint64 elements.
+
+        If bot have been encountered, we cannot safely cast to an integer
+        dtype since none is large enough to hold both types of elements.
+
+        Returns
+        -------
+        return_values : bool
+            Whether or not we should return the original input array to avoid
+            data truncation.
+        """
```

Comment:
definitly have a look at this with ``cython -a`` to see any hot spots. and profile as well.